src/bikeracks/votes.py

```diff
@@ -69,37 +69,17 @@ def submit_vote():
     # connect to db
     db = get_db()
     
-    vote_status = h.get_vote_status(db, rack_id, user_id)
-    if vote_status:
-        # updating an existing vote
-        
-        query = "UPDATE votes SET vote_type = ? WHERE rack_id = ? AND user_id = ?"
-        db.execute(query, (vote_type, rack_id, user_id))
-        db.commit()
-        # also update the vote count in bikeracks table
-        if vote_type == 1:
-            # update the downvote_count in bikeracks for this rack, decrement by 1
-            h.decrement_downvote_count(rack_id, db)
-            
-        # if the user downvoted
-        elif vote_type == -1:
-            # update the upvote_count in bikeracks for this rack, decrement by 1
-            h.decrement_upvote_count(rack_id, db)
-        
-    else:
-        # voting for the first time
-        
-        query = "INSERT INTO votes (rack_id, user_id, vote_type) VALUES (?, ?, ?)"
-        db.execute(query, (rack_id, user_id, vote_type))
-        db.commit()
-        
-    # increment the upvote or downvote count
+    query = """ INSERT INTO votes (rack_id, user_id, vote_type)
+                VALUES (?, ?, ?)
+                ON CONFLICT(rack_id, user_id) DO UPDATE SET vote_type=?"""
+    db.execute(query, (rack_id, user_id, vote_type))
+    db.commit()
+    
+    # if the vote is an upvote then delta_up is 1 and delta_down is -1
     if vote_type == 1:
-        # update the upvote_count in bikeracks for this rack, increment by 1
-        h.increment_upvote_count(rack_id, db)
-    elif vote_type == -1:
-        # update the downvote_count in bikeracks for this rack, increment by 1
-        h.increment_downvote_count(rack_id, db)
+        h.update_vote_count(db, rack_id, 1, -1)
```

Comment:
You don't know that the `down_delta` is -1 just because the `up_delta` is 1. This could be the users first vote on this rack, right?

There are three states for a rack (up, down, no vote) and you can go from any of them to any other. That's six different ways to call `update_vote_count`. Passing a single `vote_type` value from the client to the server is not enough to capture all that, you also need the previous `vote_type`.

```js
update_vote_count(.., 1, 0)  // first time voting is upvote
update_vote_count(.., 0, -1)  // first time voting is downvote
update_vote_count(.., -1, 0)  // undoing their upvote
update_vote_count(.., 0, -1)  // undoing their downvote
update_vote_count(.., 1, -1)  // changing downvote to upvote
update_vote_count(.., -1, 1)  // changing upvote to downvote
```

Reply:
'unvote' is called if a user is undoing a vote they made. Shoulld I not have unvote at all?

Reply:
I added the case where they haven't voted before.

Reply:
How does `unvote` differ from calling `update_vote_count` with `(-1, 0)` or `(0, -1)`?

Reply:
I could get rid of unvote altogether and have 6 if conditions. should I do that?

Reply:
That's up to you, but if you have two functions that do the same thing with different arguments, you should consider scrapping one of them. Usually fewer codepaths is easier to maintain.